File: library/src/main/java/me/proxer/library/api/info/InfoApi.java
package me.proxer.library.api.info;

import retrofit2.Retrofit;

/**
 * API for the Info class.
 *
 * @author Ruben Gees
 */
public final class InfoApi {

    private final InternalApi internalApi;

    /**
     * Only for internal use.
     */
    public InfoApi(final Retrofit retrofit) {
        this.internalApi = retrofit.create(InternalApi.class);
    }

    /**
     * Returns the respective endpoint.
     */
    public EntryCoreEndpoint entryCore(final String entryId) {
        return new EntryCoreEndpoint(internalApi, entryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public EntryEndpoint entry(final String entryId) {
        return new EntryEndpoint(internalApi, entryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public EpisodeInfoEndpoint episodeInfo(final String entryId) {
        return new EpisodeInfoEndpoint(internalApi, entryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public CommentsEndpoint comments(final String entryId) {
        return new CommentsEndpoint(internalApi, entryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public RelationsEndpoint relations(final String entryId) {
        return new RelationsEndpoint(internalApi, entryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public TranslatorGroupEndpoint translatorGroup(final String translatorGroupId) {
        return new TranslatorGroupEndpoint(internalApi, translatorGroupId);
    }

    /**
     * Returns the respective endpoint.
     */
    public IndustryEndpoint industry(final String industryId) {
        return new IndustryEndpoint(internalApi, industryId);
    }

    /**
     * Returns the respective endpoint.
     */
    public ModifyUserInfoEndpoint note(final String entryId) {
        return new ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.NOTE);
    }

    /**
     * Returns the respective endpoint.
     */
    public ModifyUserInfoEndpoint markAsFavorite(final String entryId) {
        return new ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.FAVORITE);
    }

    /**
     * Returns the respective endpoint.
     */
    public ModifyUserInfoEndpoint markAsFinished(final String entryId) {
        return new ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.FINISHED);
    }

    /**
     * Returns the respective endpoint.
     */
    public RecommendationsEndpoint recommendations(final String entryId) {
        return new RecommendationsEndpoint(internalApi, entryId);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/info/InfoApi.kt
package me.proxer.library.api.info

import retrofit2.Retrofit

/**
 * API for the Info class.
 *
 * @author Ruben Gees
 */
class InfoApi internal constructor(retrofit: Retrofit) {

    private val internalApi = retrofit.create(InternalApi::class.java)

    /**
     * Returns the respective endpoint.
     */
    fun entryCore(entryId: String): EntryCoreEndpoint {
        return EntryCoreEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun entry(entryId: String): EntryEndpoint {
        return EntryEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun episodeInfo(entryId: String): EpisodeInfoEndpoint {
        return EpisodeInfoEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun comments(entryId: String): CommentsEndpoint {
        return CommentsEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun relations(entryId: String): RelationsEndpoint {
        return RelationsEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun translatorGroup(translatorGroupId: String): TranslatorGroupEndpoint {
        return TranslatorGroupEndpoint(internalApi, translatorGroupId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun industry(industryId: String): IndustryEndpoint {
        return IndustryEndpoint(internalApi, industryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun note(entryId: String): ModifyUserInfoEndpoint {
        return ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.NOTE)
    }

    /**
     * Returns the respective endpoint.
     */
    fun markAsFavorite(entryId: String): ModifyUserInfoEndpoint {
        return ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.FAVORITE)
    }

    /**
     * Returns the respective endpoint.
     */
    fun markAsFinished(entryId: String): ModifyUserInfoEndpoint {
        return ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.FINISHED)
    }

    /**
     * Returns the respective endpoint.
     */
    fun subscribe(entryId: String): ModifyUserInfoEndpoint {
        return ModifyUserInfoEndpoint(internalApi, entryId, UserInfoType.SUBSCRIBE)
    }

    /**
     * Returns the respective endpoint.
     */
    fun recommendations(entryId: String): RecommendationsEndpoint {
        return RecommendationsEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun userInfo(entryId: String): MediaUserInfoEndpoint {
        return MediaUserInfoEndpoint(internalApi, entryId)
    }

    /**
     * Returns the respective endpoint.
     */
    fun forumDiscussions(entryId: String): ForumDiscussionsEndpoint {
        return ForumDiscussionsEndpoint(internalApi, entryId)
    }
}
